### PR TITLE
vkd3d: Fold contiguous WriteBufferImmediate calls

### PR DESCRIPTION
**Pull request by rlocatti-nv**
Fold contiguous WriteBufferImmediate calls into unified vkCmdUpdateBuffer invocations.
Previous to this commit, each updated DWORD would be translated into a single vkCmdUpdateBuffer call.

This helps titles using many buffer updates in their command lists.

Signed-of-by: Rodrigo Locatti rlocatti@nvidia.com

### CODE REVIEW

**Comment by doitsujin** [comment]
Which game are we talking about here?

**Comment by rlocatti-nv** [comment]
It's affecting Hitman 3.

File: libs/vkd3d/command.c [hunk 1]
```diff
@@ -9496,6 +9496,11 @@ static void STDMETHODCALLTYPE d3d12_command_list_WriteBufferImmediate(d3d12_comm
     VkDeviceSize offset;
     unsigned int i;
 
+    VkBuffer payload_buffer = VK_NULL_HANDLE;
+    size_t payload_index = 0;
+    VkDeviceSize begin_offset = 0;
+    UINT32 payload[0x40];
```
Comment (DadSchoorse) [inline libs/vkd3d/command.c:9502]: This belongs in the christmas tree above.

Comment (DadSchoorse) [reply libs/vkd3d/command.c:9500]: It's still not a christmas tree, you have to sort the lines by length. :frog:

**Comment by doitsujin** [comment]
I think there's a bit of a conflict of interest here, `WriteBufferImmediate` is **NOT** meant to be a `vkUpdateBuffer` equivalent. If games use it this way anyway then we'll have to carefully isolate the cases where this makes sense, and do it on all GPUs, vs when apps actually want `vkCmdWriteBufferMarkerAMD` sematics.

I also don't know if this is really the best solution, copying the parameter array into a scratch buffer and dispatching a compute shader that performs BDA writes may be better in case some games do lots of small disjoint updates.

In any case, I'd like to hold off on merging this and revisit this later.

**Comment by doitsujin** [comment]
Superseded by #1053.

**Pull request closed without merging**